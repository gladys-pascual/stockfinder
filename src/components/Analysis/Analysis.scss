@import "../../assets/variables.scss";
.analysis {
  padding-bottom: 10vh;
}
.analysis-title {
  @include headings;
}
.Strong-Buy {
  color: #10a57a;
}
.Buy {
  color: #6fb74e;
}
.Hold {
  color: #61a1eb;
}
.Sell {
  color: #fb9d2e;
}
.Strong-Sell {
  color: #ff526d;
}
.advice-and-chart {
  display: flex;
  align-items: center;
  @media screen and (max-width: 700px) {
    flex-direction: column;
  }
  .advice {
    padding: 5%;
    .advice-date,
    .disclaimer {
      font-weight: 300;
    }
    .advice-heading {
      @include headings;
      font-weight: 500;
    }
    .disclaimer {
      font-style: italic;
    }
  }
}
.analysis-bar-chat-wrapper {
  padding-right: 5%;
  @media screen and (max-width: 700px) {
    padding-right: 0;
  }
  .analysis-bar-chart {
    height: 450px;
    width: 60vw;
    border: 1px darkgrey solid;
    border-radius: 4px;
    margin: 12px;
    @media screen and (max-width: 700px) {
      width: 90vw;
    }
  }
}
