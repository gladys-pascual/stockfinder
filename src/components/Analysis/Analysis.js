import "./Analysis.scss";
import PropTypes from "prop-types";
import { useEffect, useState } from "react";
import axios from "axios";
import LoadingSearchAPI from "../Loading/LoadingSearchAPI";
import AnalysisBarChart from "./AnalysisBarChart";

const Analysis = ({ symbol, companyName }) => {
  const [analysisData, setAnalysisData] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    async function fetchAnalysisData() {
      try {
        const result = await axios.get(
          `https://finnhub.io/api/v1/stock/recommendation?symbol=${symbol}&token=${process.env.REACT_APP_API_KEY}`
        );
        setAnalysisData(result.data[0]);
        setLoading(false);
      } catch (err) {
        console.log("Error fetching and parsing data", err);
        setLoading(false);
      }
    }
    fetchAnalysisData();
  }, [symbol]);

  if (loading) {
    return <LoadingSearchAPI />;
  }

  const period = analysisData.period;
  const formatDate = (date) => {
    const splitDate = date.split("");

    const year = splitDate.slice(0, 4).join("");
    const day = splitDate.slice(8, 10).join("");
    const monthNum = splitDate.slice(5, 7).join("");
    let month;
    if (monthNum === "01") {
      month = "Jan";
    } else if (monthNum === "02") {
      month = "Feb";
    } else if (monthNum === "03") {
      month = "Mar";
    } else if (monthNum === "04") {
      month = "Apr";
    } else if (monthNum === "05") {
      month = "May";
    } else if (monthNum === "06") {
      month = "Jun";
    } else if (monthNum === "07") {
      month = "Jul";
    } else if (monthNum === "08") {
      month = "Aug";
    } else if (monthNum === "09") {
      month = "Sep";
    } else if (monthNum === "10") {
      month = "Oct";
    } else if (monthNum === "11") {
      month = "Nov";
    } else if (monthNum === "12") {
      month = "Dec";
    }

    return `${month} ${day} ${year}`;
  };

  const strongBuy = analysisData.strongBuy;
  const buy = analysisData.buy;
  const hold = analysisData.hold;
  const sell = analysisData.sell;
  const strongSell = analysisData.strongSell;
  const total = strongBuy + buy + hold + sell + strongSell;

  const recommendationData = {
    "Strong Buy": strongBuy,
    Buy: buy,
    Hold: hold,
    Sell: sell,
    "Strong Sell": strongSell,
  };

  let maxVote = 0;
  let maxVoteKey;
  let maxVoteKeyClassName;

  for (const recommendation in recommendationData) {
    if (recommendationData[recommendation] > maxVote) {
      maxVote = recommendationData[recommendation];
      maxVoteKey = recommendation;
      maxVoteKeyClassName = recommendation.replace(" ", "-");
    }
  }

  return (
    <section>
      <h1 className="analysis-title">
        {`Analysis for ${companyName}`}{" "}
        <i className="fas fa-chart-bar">
          <span className="sr-only">Bar chart icon</span>
        </i>
      </h1>
      <div className="advice-and-chart">
        <div className="advice">
          <p className="advice-date">{`Based on ${total} analyst, as of ${formatDate(
            period
          )}:`}</p>
          <h1 className={`advice-heading ${maxVoteKeyClassName}`}>
            {" "}
            {maxVoteKey}
          </h1>
          <p className="disclaimer">
            Disclaimer: The analysis information is for reference only and does
            not constitute an investment recommendation.
          </p>
        </div>
        <div className="analysis-bar-chat-wrapper">
          <AnalysisBarChart
            strongBuy={strongBuy}
            buy={buy}
            hold={hold}
            sell={sell}
            strongSell={strongSell}
            total={total}
          />
        </div>
      </div>
    </section>
  );
};

Analysis.propTypes = {
  symbol: PropTypes.string.isRequired,
  companyName: PropTypes.string.isRequired,
};
export default Analysis;
