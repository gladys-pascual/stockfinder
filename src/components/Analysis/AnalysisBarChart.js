import PropTypes from "prop-types";
import { ResponsiveBar } from "@nivo/bar";
import "./Analysis.scss";

const AnalysisBarChart = ({
  strongBuy,
  buy,
  hold,
  sell,
  strongSell,
  total,
}) => {
  const strongBuyPercentage = +((strongBuy / total) * 100).toFixed(2);
  const buyPercentage = +((buy / total) * 100).toFixed(2);
  const holdPercentage = +((hold / total) * 100).toFixed(2);
  const sellPercentage = +((sell / total) * 100).toFixed(2);
  const strongSellPercentage = +((strongSell / total) * 100).toFixed(2);

  const data = [
    {
      analysis: "Strong Buy",
      "Strong Buy": strongBuyPercentage,
    },
    {
      analysis: "Buy",
      Buy: buyPercentage,
    },
    {
      analysis: "Hold",
      Hold: holdPercentage,
    },
    {
      analysis: "Sell",
      Sell: sellPercentage,
    },
    {
      analysis: "Strong Sell",
      "Strong Sell": strongSellPercentage,
    },
  ];

  return (
    <section className="analysis-bar-chart">
      <ResponsiveBar
        data={data}
        indexBy="analysis"
        keys={["Strong Buy", "Buy", "Hold", "Sell", "Strong Sell"]}
        margin={{ top: 50, right: 0, bottom: 50, left: 0 }}
        padding={0.3}
        colors={["#FF526D", "#FB9D2E", "#A7B0BA", "#6FB74E", "#10A57A"]}
        label={(d) => `${d.value}%`}
        enableGridY={false}
        axisTop={null}
        axisRight={null}
        axisBottom={{
          tickSize: 10,
          tickPadding: 5,
          tickRotation: 0,
          legendPosition: "middle",
          legendOffset: 32,
        }}
        axisLeft={null}
        labelSkipWidth={12}
        labelSkipHeight={12}
        labelTextColor={{ from: "color", modifiers: [["darker", 1.6]] }}
        isInteractive={false}
        animate={true}
        motionStiffness={90}
        motionDamping={15}
      />
    </section>
  );
};
AnalysisBarChart.propTypes = {
  strongBuy: PropTypes.number.isRequired,
  buy: PropTypes.number.isRequired,
  hold: PropTypes.number.isRequired,
  sell: PropTypes.number.isRequired,
  strongSell: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
};

export default AnalysisBarChart;
