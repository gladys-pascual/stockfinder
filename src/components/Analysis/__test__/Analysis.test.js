import { render } from "@testing-library/react";
import Analysis from "../Analysis";

test("render Analysis component as expectetd", () => {
  const props = {
    symbol: "AAPL",
    companyName: "Apple Inc",
  };

  const { container } = render(<Analysis {...props} />);

  expect(container).toMatchSnapshot();
});
