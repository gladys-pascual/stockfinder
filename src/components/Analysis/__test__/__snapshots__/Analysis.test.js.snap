// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Analysis component as expectetd 1`] = `
<div>
  <div
    class="loading-search-api"
  >
    <span
      class="css-0"
    >
      <span
        class="css-158729u"
      />
      <span
        class="css-jyyt77"
      />
      <span
        class="css-uj62sn"
      />
    </span>
  </div>
</div>
`;
